astropy/wcs/utils.py

```diff
@@ -632,3 +632,203 @@ def pixel_to_skycoord(xp, yp, wcs, origin=0, mode='all', cls=None):
+    projection : None or str
+        Three-letter projection code. Defaults to TAN.
+    proj_point: `~astropy.coordinates.SkyCoord` or str
+        Celestial coordinates of projection point (lat, lon). If 'center', the geometric
+        center of input coordinates will be used for the projection. If None, and input
+        wcs with CRVAL set must be passed in.
+    order : None, int, or tuple of ints
+        A order, B order, respectivley, for SIP polynomial to be fit, or single integer
+        for both. Must be provided if mode is 'sip'.
+
+    Returns
+    -------
+    wcs : `~astropy.wcs.WCS`
+        WCS object with best fit coefficients given the matched set of X, Y pixel and
+        sky positions of sources.
+
+    Notes
+    -----
+    Please pay careful attention to the logic that applies when both an input WCS with
+    keywords populated is passed in. For example, if no 'CUNIT' is set in this input
+    WCS, the units of the CRVAL etc. are assumed to be the same as the input Skycoord.
+    Additionally, if 'RADESYS' is not set in the input WCS, this will be taken from the
+    Skycoord as well. """
+
+    from scipy.optimize import least_squares
+    from .wcs import Sip
+    import copy
+
+    lon, lat = coords.data.lon.deg, coords.data.lat.deg
+    mode = mode.lower()
+    if mode not in ("wcs", "all"):
+        raise ValueError("mode must be 'wcs' or 'all'")
+    if inwcs is not None:
+        inwcs = copy.deepcopy(inwcs)
+    if projection is None:
+        if (inwcs is None) or ('' in inwcs.wcs.ctype):
+            raise ValueError("Must provide projection type or input WCS with CTYPE.")
+        projection = inwcs.wcs.ctype[0].replace('-SIP','').split('-')[-1]
+
+    # template wcs
+    wcs = celestial_frame_to_wcs(frame=coords.frame, projection=projection)
+
+    # Determine CRVAL from input, and
+    close = lambda l, p: p[np.where(np.abs(l) == min(np.abs(l)))[0][0]]
+    if str(proj_point) == 'center': #use center of input points
+        wcs.wcs.crval = ((max(lon)+min(lon))/2.,(max(lat)+min(lat))/2.)
+        wcs.wcs.crpix = ((max(xp)+min(xp))/2.,(max(yp)+min(yp))/2.) #initial guess
+    elif (proj_point is None) and (inwcs is None):
+        raise ValueError("Must give proj_point as argument or as CRVAL in input wcs.")
+    elif proj_point is not None: # convert units + rough initial guess for crpix for fit
+        lon_u, lat_u = u.Unit(coords.data.lon.unit), u.Unit(coords.data.lat.unit)
+        wcs.wcs.crval = (proj_point[0]*lon_u.to(u.deg), proj_point[1]*lat_u.to(u.deg))
+        wcs.wcs.crpix = (close(lon-wcs.wcs.crval[0],xp),close(lon-wcs.wcs.crval[1],yp))
+
+    if inwcs is not None:
+        if inwcs.wcs.radesys == '': # no frame specified, use Skycoord's frame (warn???)
+            inwcs.wcs.radesys = coords.frame.name
+        wcs.wcs.radesys = inwcs.wcs.radesys
+        coords.transform_to(inwcs.wcs.radesys.lower()) #work in inwcs units
```

Comment:
Two points about the lines above:

- `wcs.radesys` exists only for informational purpose, it's not used in any computations. The reference frame is actually coded in `CTYPE`. Assigning `radesys` like that does not sync the two objects because `CTYPEs` may encode a different reference frame.

`utils.wcs_to_celestial_frame` and `celestial_frame_to_wcs` can be used.